components/eamxx/src/physics/cosp/eamxx_cosp.cpp

```diff
@@ -78,8 +78,9 @@ void Cosp::set_grids(const std::shared_ptr<const GridsManager> grids_manager)
 // =========================================================================================
 void Cosp::initialize_impl (const RunType /* run_type */)
 {
-  // Determine how often to call COSP, specified as number of atm steps (>0) or number of hours (<0)
+  // Determine how often to call COSP; units can be steps or hours
   m_cosp_frequency = m_params.get<Int>("cosp_frequency", 1);
+  m_cosp_frequency_units = m_params.get<std::string>("cosp_frequency_units", "steps");
```

Comment:
I would move this in the constructor, and move the check that you added in run_impl (checking that units is either "steps" or "hours") in the constructor too.

The reason is that initialize_impl for atm processes is called toward the end of the init sequence. If the user specified invalid units, we would like to catch that asap. Esp for ne1024 runs, it may save quite a few core hours.

Reply:
Is there any reason not to put ALL the param reads into the constructor then?